File: src/commonTest/kotlin/com/scoreo/application/AddPlayerUseCaseTest.kt
package com.scoreo.application

import com.scoreo.domain.DomainError
import com.scoreo.infrastructure.InMemoryPlayerRepository
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertNotEquals
import kotlin.test.assertTrue

class AddPlayerUseCaseTest {

    @Test
    fun `adds player to repository`() {
        val repo = InMemoryPlayerRepository()
        val useCase = AddPlayerUseCase(repo)

        useCase("Alice")

        assertEquals(1, repo.getAll().size)
        assertEquals("Alice", repo.getAll().first().name)
    }

    @Test
    fun `trims whitespace from name`() {
        val repo = InMemoryPlayerRepository()
        val useCase = AddPlayerUseCase(repo)

        useCase("  Bob  ")

        assertEquals("Bob", repo.getAll().first().name)
    }

    @Test
    fun `each player gets a unique id`() {
        val repo = InMemoryPlayerRepository()
        val useCase = AddPlayerUseCase(repo)

        useCase("Alice")
        useCase("Bob")

        val ids = repo.getAll().map { it.id }
        assertNotEquals(ids[0], ids[1])
    }

    @Test
    fun `multiple players are all saved`() {
        val repo = InMemoryPlayerRepository()
        val useCase = AddPlayerUseCase(repo)

        useCase("Alice")
        useCase("Bob")
        useCase("Charlie")

        assertEquals(3, repo.getAll().size)
    }

    @Test
    fun `rejects name exceeding 50 characters`() {
        val repo = InMemoryPlayerRepository()
        val useCase = AddPlayerUseCase(repo)
        val longName = "A".repeat(51)
        val exception = assertFailsWith<DomainError.Validation> {
            useCase(longName)
        }
        assertTrue(exception.message!!.contains("50"))
    }

    @Test
    fun `accepts name of exactly 50 characters`() {
        val repo = InMemoryPlayerRepository()
        val useCase = AddPlayerUseCase(repo)
        val name = "A".repeat(50)
        val player = useCase(name)
        assertEquals(50, player.name.length)
    }
}

File: apps/scoreo/src/application/addPlayerUseCase.test.ts
import { describe, expect, it } from 'vitest'
import { ValidationError } from '../domain/model/errors'
import { InMemoryPlayerRepository } from '../infrastructure/testing/inMemoryPlayerRepository'
import { AddPlayerUseCase } from './addPlayerUseCase'

describe('AddPlayerUseCase', () => {
  it('adds player to repository', () => {
    const repo = new InMemoryPlayerRepository()
    const useCase = new AddPlayerUseCase(repo)

    useCase.invoke('Alice')

    expect(repo.getAll()).toHaveLength(1)
    expect(repo.getAll()[0].name).toBe('Alice')
  })

  it('trims whitespace from name', () => {
    const repo = new InMemoryPlayerRepository()
    const useCase = new AddPlayerUseCase(repo)

    useCase.invoke('  Bob  ')

    expect(repo.getAll()[0].name).toBe('Bob')
  })

  it('each player gets a unique id', () => {
    const repo = new InMemoryPlayerRepository()
    const useCase = new AddPlayerUseCase(repo)

    useCase.invoke('Alice')
    useCase.invoke('Bob')

    const ids = repo.getAll().map((p) => p.id)
    expect(ids[0]).not.toBe(ids[1])
  })

  it('multiple players are all saved', () => {
    const repo = new InMemoryPlayerRepository()
    const useCase = new AddPlayerUseCase(repo)

    useCase.invoke('Alice')
    useCase.invoke('Bob')
    useCase.invoke('Charlie')

    expect(repo.getAll()).toHaveLength(3)
  })

  it('rejects name exceeding 50 characters', () => {
    const repo = new InMemoryPlayerRepository()
    const useCase = new AddPlayerUseCase(repo)
    const longName = 'A'.repeat(51)

    let error: unknown
    try {
      useCase.invoke(longName)
    } catch (e) {
      error = e
    }
    expect(error).toBeInstanceOf(ValidationError)
    expect((error as ValidationError).message).toContain('50')
  })

  it('accepts name of exactly 50 characters', () => {
    const repo = new InMemoryPlayerRepository()
    const useCase = new AddPlayerUseCase(repo)
    const name = 'A'.repeat(50)

    const player = useCase.invoke(name)

    expect(player.name).toHaveLength(50)
  })
})
